File: 516.最长回文子序列.java
import java.util.Vector;

/*
 * @Description: 
 * @Author: yangyuxiang
 * @Date: 2021-04-27 07:46:17
 * @LastEditors: yangyuxiang
 * @LastEditTime: 2021-04-27 07:58:48
 * @FilePath: /leetcode/516.最长回文子序列.java
 */
/*
 * @lc app=leetcode.cn id=516 lang=java
 *
 * [516] 最长回文子序列
 */

// @lc code=start
class LongestPalindromeSubseq {
    public int longestPalindromeSubseq(String s) {
        if (s.length() <= 1) { 
            return s.length(); 
        }
        int[][] dp = new int[s.length()][s.length()];
        for (int i = s.length()-1; i >= 0; i--) {
            for (int j = i; j < s.length(); j++) {
                if (i == j) {
                    dp[i][j] = 1;
                    continue;
                }
                if (s.charAt(i) == s.charAt(j)) {
                    dp[i][j] = dp[i+1][j-1] + 2;
                } else {
                    dp[i][j] = Math.max(dp[i][j-1], dp[i+1][j]);
                }
                
            }
        }

        return dp[0][s.length()-1];


    }
}
// @lc code=end

File: 516.最长回文子序列.py
#
# @lc app=leetcode.cn id=516 lang=python
#
# [516] 最长回文子序列
#

# @lc code=start
class Solution(object):
    def longestPalindromeSubseq0(self, s):
        """
        :type s: str
        :rtype: int
        解法：求s和s反转后的字符串的最长公共子序列
        """
        if len(s) <= 1:
            return len(s)
        
        n = len(s)
        s1 = s
        s2 = "".join(list(s)[::-1])
        dp = [[0 for _ in range(n+1)] for _ in range(n+1)]
        for i in range(1, n+1):
            for j in range(1, n+1):
                if s1[i-1] == s2[j-1]:
                    dp[i][j] = dp[i-1][j-1] + 1
                else:
                    dp[i][j] = max(dp[i-1][j], dp[i][j-1])

        return dp[-1][-1]

    def longestPalindromeSubseq(self, s):
        """
        :type s: str
        :rtype: int
        解法：dp, 直接求最长回文子序列的长度，dp[i][j]表示s[i:j]的最长回文子序列的长度
        """
        if len(s) <= 1:
            return len(s)
        
        n = len(s)
        dp = [[0 for _ in range(n)] for _ in range(n)]
        for i in range(n-1, -1, -1):
            for j in range(i, n):
                if j == i:
                    dp[i][j] = 1
                    continue
                if s[i] == s[j]:
                    dp[i][j] = dp[i+1][j-1] + 2
                else:
                    dp[i][j] = max(dp[i+1][j], dp[i][j-1])

        return dp[0][n-1]


if __name__ == '__main__':
    s = "cbbd"
    print(Solution().longestPalindromeSubseq(s))

# @lc code=end

